### DISCUSSION TITLE
Stage 2 of tests

### DISCUSSION THREAD

**Discussion by haipinglu** [pykale/pykale#140] [Testing]
Having achieved >70% coverage, we reached the first milestone (let's call it Stage 1) and the codecov badge is on. This discussion compiles the remaining issues about tests, in Stage 2 of test development. [See current coverage status.](https://codecov.io/gh/pykale/pykale/list/main/kale).

**_Important rule: add new tests whenever you fix a bug, that catches the previous bug._**

- **_Data for test efficiency_**: So far, we used some real data for tests. Some are necessary (e.g., `loaddata` API) but others are not (e.g. `domain_adapter`). It would be better to construct smaller random datasets or use small subsets of real data to replace real data used as much as possible to improve the testing time needed, e.g., replacing USPS in domain adaptation tests! This is not critical at the moment since the tests are done on servers via CI and free.
- **_`tests` directory structure_**
  -  Consider whether to remove `test_data` folder entirely and move to `pykale/data` or keep some small data files there.
  - Delete unused (but almost fully covered) test files in tests. Keep on those necessary or with a low coverage.
- **_Depth of tests_**: Some tests are really basic, e.g. just variable type in `digit_access`. In future, test more properties of the data loader. Of course, this may not be uncommon in ML/DL libraries, e.g. [torchvision has a coverage of 80% as of 2021.05.04](https://codecov.io/gh/pytorch/vision/list/master/torchvision) but this does not reduce its popularity and usefulness. It is more a trade-off (priority) between spending the limited amount of efforts on the core API or tests. Of course, the ideal case is for these to be aligned.
  - From @XianyuanLiu: We may add more detailed tests _in the future_ for `sampler` and `losses`. The current one may be too high-level. We can create the simplest model and data loaders like [`boring_model`](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/helpers/boring_model.py#L63) to fix the loss output for the test. We can also test our sampler output size by following [PyTorch sampler test](https://github.com/pytorch/vision/blob/master/test/test_datasets_samplers.py).
TorchVision test: https://github.com/pytorch/vision/tree/master/test
PyTorch-Lightning test:  https://github.com/PyTorchLightning/pytorch-lightning/tree/master/tests

Being aware of the needs above helps us prioritize our tasks, particularly research works and paper deadlines, and manage the project properly to allocate/distribute resources, e.g., @bobturneruk,  students, and other potential contributors. We will discuss the above in group meetings and keep it updated.

**Comment by bobturneruk** [comment]
Good points. I'll invite some of my colleagues to venture their thoughts.

**Reply by bobturneruk** [reply]
From @Robadob, thanks:

> Yeah coverage suffers from Goodhart's law.

> Ideal world, tests are all small and self-contained. So when one fails you know the exact problem straight away. But real world that's rarely possible and there will be overlap. In FGPU2, we generally try and and cover any different configurations of each function (or functionality), especially any edge cases. We haven't managed to find a way to do coverage across CUDA, so no idea how much we're missing. Only important rule would be to add new tests whenever you fix a bug, that catches the previous bug. (edited) 

> I'm not even too sure about how pykale works, so some testing strategies might be futile.
E.g. parsing code is notoriously hard to test thoroughly. Google's project zero has been fuzzing ffmpeg (widely used video encode/decode lib) for a decade, and is still finding serious bugs. One of the libav devs once said at a conference, that it will never be bug free.

**Reply by bobturneruk** [reply]
Something else to bear in mind is that when bugs are located and fixed, tests should be added that would catch the bug.

**Reply by haipinglu** [reply]
> Something else to bear in mind is that when bugs are located and fixed, tests should be added that would catch the bug.

Yes, this is a good point and I've added at the top post in bold and please see PR #142.

**Comment by bobturneruk** [comment]
Some very detailed looking stuff here https://arxiv.org/abs/1906.10742

**Reply by haipinglu** [reply]
> Some very detailed looking stuff here https://arxiv.org/abs/1906.10742

I know the last author and they are experts in testing. We have no intention to switch our research focus to testing. Would you be able to dig out from the 37 pages what are potentially useful to us and maybe give us a presentation and work out some examples (for pykale ideally)?

**Reply by bobturneruk** [reply]
Yes - it's quite academic / theoretical in places and my approach is much more engineering / practical. I'll see if it's feasible to put together some actionable stuff for PyKale.

**Reply by haipinglu** [reply]
> Yes - it's quite academic / theoretical in places and my approach is much more engineering / practical. I'll see if it's feasible to put together some actionable stuff for PyKale.

To be feasible, we should take a look and learn from how pytorch libraries are doing (see top suggestions). Otherwise, the implementation cost could be very high and become impractical.

**Comment by bobturneruk** [comment]
This might be worth a look, if only to see what it's doing https://github.com/suriyadeepan/torchtest

**Reply by haipinglu** [reply]
> This might be worth a look, if only to see what it's doing https://github.com/suriyadeepan/torchtest

There have been two years since the last update and most of code there are outdated. Would you be interested to "upgrade" it and test it out?

**Reply by bobturneruk** [reply]
Sure, if we think there's value in what it's supposed to do.

**Comment by xianyuanliu** [comment]
Pytorch-ada has some testing examples which can help us improve. (sampler, dataset_access and multi_domain)

https://github.com/criteo-research/pytorch-ada/tree/master/tests/datasets
